src/jit/compiler.cpp

```diff
@@ -11097,3 +11097,831 @@ bool Compiler::killGCRefs(GenTree* tree)
+    m_VariableLifeBarrier = allocator.allocate<LiveRangeDumper>(1);
+
+    size_t m_VariableLifeBarrierSize = sizeof(*m_VariableLifeBarrier);
+    memset(m_VariableLifeBarrier, 0, m_VariableLifeBarrierSize);
+
+    new (m_VariableLifeBarrier, jitstd::placement_t()) LiveRangeDumper(m_VariableLiveRanges);
+#endif // DEBUG
+}
+
+VariableLiveDescriptor::~VariableLiveDescriptor()
+{
+    noway_assert(m_VariableLiveRanges != nullptr);
+
+    m_VariableLiveRanges->~list();
+
+#ifdef DEBUG
+    m_VariableLifeBarrier->~LiveRangeDumper();
+#endif // DEBUG
+}
+
+//------------------------------------------------------------------------
+// hasVariableLiveRangeOpen: Return true if the variable is still alive,
+//  false in other case.
+//
+bool VariableLiveDescriptor::hasVariableLiveRangeOpen() const
+{
+    noway_assert(m_VariableLiveRanges != nullptr);
+
+    return !m_VariableLiveRanges->empty() && !m_VariableLiveRanges->back().m_EndEmitLocation.Valid();
+}
+
+//------------------------------------------------------------------------
+// getAmountLiveRanges: Return amount of variable homes reported for this
+//  variable.
+//
+// Return Value:
+//  A boolean.
+//
+size_t VariableLiveDescriptor::getAmountLiveRanges() const
+{
+    size_t result = 0;
+    if (m_VariableLiveRanges != nullptr)
+    {
+        result = m_VariableLiveRanges->size();
+    }
+    return result;
+}
+
+//------------------------------------------------------------------------
+// getAmountLiveRanges: Return the list of variable homes for this variable.
+//
+// Return Value:
+//  A const LiveRangeList* pointing to the first variable home if it has
+//  any or the end of the list in other case.
+//
+const LiveRangeList* VariableLiveDescriptor::getLiveRanges() const
+{
+    return m_VariableLiveRanges;
+}
+
+//------------------------------------------------------------------------
+// startLiveRangeFromEmitter: Report this variable as being born in "varHome"
+//  since the instruction where "_emitter" is located.
+//
+// Arguments:
+//  varHome  - the home of the variable.
+//  _emitter - an emitter* instance located at the first instruction from
+//   where "varHome" becomes valid.
+//
+// Assumptions:
+//  This variable is being born so it should be dead.
+//
+// Notes:
+//  The position of "_emitter" matters to ensure intervals inclusive of the
+//  beginning and exclusive of the end.
+//
+void VariableLiveDescriptor::startLiveRangeFromEmitter(CodeGenInterface::siVarLoc varHome, emitter* _emitter) const
+{
+    noway_assert(_emitter != nullptr);
+
+    // Is the first "VariableLiveRange" or the previous one has been closed so its "m_EndEmitLocation" is valid
+    noway_assert(m_VariableLiveRanges->empty() || m_VariableLiveRanges->back().m_EndEmitLocation.Valid());
+
+    // Creates new live range with invalid end
+    m_VariableLiveRanges->emplace_back(varHome, emitLocation(), emitLocation());
+    m_VariableLiveRanges->back().m_StartEmitLocation.CaptureLocation(_emitter);
```

Comment:
Seems kind of strange to first add the location to the list and then update it. I'd rather have:
```C++
    emitLocation startLocation;
    startLocation.CaptureLocation(emitter);
    m_VariableLiveRanges.emplace_back(varHome, startLocation, emitLocation());
```
or maybe update emitLocation constructors such that you can write
```C++
    m_VariableLiveRanges.emplace_back(varHome, emitLocation(emitter), emitLocation());
```